_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,416 @@
+
+The number of records acquired for a partition by consumers in a shared group is also limited. Once this limit is reached, fetching of records from the shared group temporarily pauses until the number of acquired records reduces. The limit can be configured with the  broker configuration, <b>group.share.partition.max.record.locks</b> which is set to 200 by default.
+
+Records transition through different states when being fetched depending on the actions taken by consumers of a shared group. Consumer can:
+- Acknowledge once the record is successfully processed
+- Release the record, to make it available for another delivery attempt
+- Reject the record, if it is unprocessable and the record should not be available for another delivery attempt
+
+If none of the above is done by the consumer e.g. the consumer is still in the middle of processing the record and is taking a long time, the lock will be automatically released once the lock duration is elapsed. 
+
+
+### Delivery State
+
+![Delivery state](/assets/images/posts/2025-06-24-kafka-queue-02.png)
+
+Delivery count for each record is tracked and gets incremented every time a consumer acquires the record. There is a limit on the number of times records can be retried to avoid getting stuck trying to deliver unprocessable records indefinitely. This limit also can be reconfigured with broker configuration, <b>group.share.delivery.count.limit</b> with default value of 5.
+
+The delivery count determines which state the record should transition into later on. If the consumer releases the record or the lock duration has lapsed, it will go back into <b>Available</b> state as long as the delivery count has not reached the limit. 
+The record will go into <b>Acknowledged</b> state if it has been successfully processed by the consumer. If the consumer rejects the record or its delivery count has reached the limit, the record will go into <b>Archived</b> state. 
+
+In the traditional message queue, once a record is fetched by a consumer, it is no longer in the queue. You can think of <b>Archived</b> state, as the records that are taken off the queue and no longer available. 
+
+### Share Group State
+
+The share group persists state management for subscribed partitions, by storing it in an internal topic called, <code>__share_group_state</code>. This topic stores state of share groups, for example, the partitions they are subscribed to and the state of those partitions such as their fetched start offset (SPSO) and the end offset (SPEO) as well as the delivery state and delivery count of the records on the partitions.
+
+We can take a look at how this information is stored in the <code>__share_group_state</code> topic by using the console consumer tool:
+```
+./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic __shared_group_state | jq .
```

Comment:
This seems a bit disconnected as users cannot run this with Strimzi. I wonder if it would be possible to move it only behind the chapter where you have the cluster and change the command to allow users to try this themself.

I also wonder how did the shared group got there -> wouldn't it show up there only after being used?